client/java/src/main/java/com/linecorp/centraldogma/client/TransformingWatcher.java

```diff
@@ -18,91 +18,148 @@
-        return parent.initialValueFuture().thenApply(ignored -> latest());
+        return initialValueFuture;
     }
 
     @Override
     public Latest<U> latest() {
-        if (!closed) {
-            final Latest<T> latestParent = parent.latest();
-            final U transformedValue = transformer.apply(latestParent.value());
-            transformedLatest = new Latest<>(latestParent.revision(), transformedValue);
+        final Latest<U> mappedLatest = this.mappedLatest;
+        if (mappedLatest == null) {
+            throw new IllegalStateException("value not available yet");
         }
-        return transformedLatest;
+        return mappedLatest;
     }
 
     @Override
     public void close() {
         closed = true;
+        if (!initialValueFuture.isDone()) {
+            initialValueFuture.cancel(false);
+        }
         // do nothing else. We don't own the parent's lifecycle
     }
 
     @Override
     public void watch(BiConsumer<? super Revision, ? super U> listener) {
-        requireNonNull(listener, "listener");
-        parent.watch(transform(listener));
+        watch(listener, parent.watchScheduler());
     }
 
     @Override
     public void watch(BiConsumer<? super Revision, ? super U> listener, Executor executor) {
         requireNonNull(listener, "listener");
         requireNonNull(executor, "executor");
-        parent.watch(transform(listener), executor);
-    }
+        updateListeners.add(Maps.immutableEntry(listener, executor));
 
-    private BiConsumer<Revision, T> transform(BiConsumer<? super Revision, ? super U> listener) {
-        return (revision, value) -> {
-            if (closed) {
-                return;
-            }
-            final U transformedValue = transformer.apply(value);
-            final boolean changed;
-            if (firstWatch) {
-                firstWatch = false;
-                changed = true;
-            } else {
-                changed = !transformedLatest.value().equals(transformedValue);
-            }
-            if (changed) {
-                transformedLatest = new Latest<>(revision, transformedValue);
-                listener.accept(revision, transformedValue);
-            } // else, transformed value has not changed
-        };
+        final Latest<U> mappedLatest = this.mappedLatest;
+        if (mappedLatest != null) {
+            // There's a chance that listener.accept(...) is called twice for the same value
+            // if this watch method is called:
+            // - after "mappedLatest = newLatest;" is invoked.
+            // - and before notifyListener() is called.
+            // However, it's such a rare case and we usually call `watch` method after creating a Watcher,
+            // which means mappedLatest is probably not set yet, so we don't use a lock to guarantee
+            // the atomicity.
+            executor.execute(() -> {
+                listener.accept(mappedLatest.revision(), mappedLatest.value());
```

Comment:
How about reading `mappedLatest` again? The value might be changed before starting the `Runnable` task.
```suggestion
                final Latest<U> mappedLatest0 = this.mappedLatest;
                listener.accept(mappedLatest0.revision(), mappedLatest0.value());
```